libs/wire-api/src/Wire/API/Conversation/Action.hs

```diff
@@ -14,74 +14,179 @@
-  | ConversationActionDelete
-  deriving stock (Eq, Show, Generic)
-  deriving (Arbitrary) via (GenericUniform ConversationAction)
-  deriving (ToJSON, FromJSON) via (CustomEncoded ConversationAction)
+
+data ConversationActionTag
+  = ConversationJoinTag
+  | ConversationLeaveTag
+  | ConversationRemoveMembersTag
+  | ConversationMemberUpdateTag
+  | ConversationDeleteTag
+  | ConversationRenameTag
+  | ConversationMessageTimerUpdateTag
+  | ConversationReceiptModeUpdateTag
+  | ConversationAccessDataTag
+  deriving (Show, Eq, Generic, Bounded, Enum)
+
+instance Arbitrary ConversationActionTag where
+  arbitrary = elements [minBound .. maxBound]
+
+instance ToSchema ConversationActionTag where
+  schema =
+    enum @Text "ConversationActionTag" $
+      mconcat
+        [ element "ConversationJoinTag" ConversationJoinTag,
+          element "ConversationLeaveTag" ConversationLeaveTag,
+          element "ConversationRemoveMembersTag" ConversationRemoveMembersTag,
+          element "ConversationMemberUpdateTag" ConversationMemberUpdateTag,
+          element "ConversationDeleteTag" ConversationDeleteTag,
+          element "ConversationRenameTag" ConversationRenameTag,
+          element "ConversationMessageTimerUpdateTag" ConversationMessageTimerUpdateTag,
+          element "ConversationReceiptModeUpdateTag" ConversationReceiptModeUpdateTag,
+          element "ConversationAccessDataTag" ConversationAccessDataTag
+        ]
+
+instance ToJSON ConversationActionTag where
+  toJSON = schemaToJSON
+
+instance FromJSON ConversationActionTag where
+  parseJSON = schemaParseJSON
+
+$(genSingletons [''ConversationActionTag])
+
+-- | We use this type family instead of a sum type to be able to define
+-- individual effects per conversation action. See 'HasConversationActionEffects'.
+type family ConversationAction (tag :: ConversationActionTag) :: * where
+  ConversationAction 'ConversationJoinTag = ConversationJoin
+  ConversationAction 'ConversationLeaveTag = ConversationLeave
+  ConversationAction 'ConversationMemberUpdateTag = ConversationMemberUpdate
+  ConversationAction 'ConversationDeleteTag = ConversationDelete
+  ConversationAction 'ConversationRenameTag = ConversationRename
+  ConversationAction 'ConversationMessageTimerUpdateTag = ConversationMessageTimerUpdate
+  ConversationAction 'ConversationReceiptModeUpdateTag = ConversationReceiptModeUpdate
+  ConversationAction 'ConversationAccessDataTag = ConversationAccessData
+  ConversationAction 'ConversationRemoveMembersTag = ConversationRemoveMembers
+
+data SomeConversationAction where
+  SomeConversationAction :: Sing tag -> ConversationAction tag -> SomeConversationAction
+
+instance Show SomeConversationAction where
+  show (SomeConversationAction tag action) =
+    $(sCases ''ConversationActionTag [|tag|] [|show action|])
+
+instance Eq SomeConversationAction where
+  (SomeConversationAction tag1 action1) == (SomeConversationAction tag2 action2) =
+    case (tag1, tag2) of
```

Comment:
This can be improved using decidable equality of tags:
```haskell
instance Eq SomeConversationAction where
  (SomeConversationAction tag1 action1) == (SomeConversationAction tag2 action2) =
    case tag1 %~ tag2 of
      Proved Refl -> $(sCases ''ConversationActionTag [|tag1|] [|action1 == action2|])
      Disproved _ -> False
```